### PR TITLE
Add checks for old cluster features in rolling upgrade tests

### PR DESCRIPTION
**Pull request by ldematte**
Add the ability to test for the original/old cluster features during a rolling upgrade

### CODE REVIEW

**Comment by ldematte** [comment]
@elasticmachine update branch

**Comment by ldematte** [comment]
@elasticmachine rerun elasticsearch-ci/part-3

File: qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/ParameterizedRollingUpgradeTestCase.java [hunk 1]
```diff
@@ -78,6 +80,13 @@ protected ParameterizedRollingUpgradeTestCase(@Name("upgradedNodes") int upgrade
         this.requestedUpgradedNodes = upgradedNodes;
     }
 
+    @Before
```
Comment (thecoop) [inline qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/ParameterizedRollingUpgradeTestCase.java:83]: The clear is an `AfterClass`. So should this be `BeforeClass`?

File: test/framework/src/main/java/org/elasticsearch/test/rest/TestFeatureService.java [hunk 2]
```diff
@@ -8,8 +8,24 @@
...
 
+import java.util.Set;
+
 public interface TestFeatureService {
     boolean clusterHasFeature(String featureId);
 
-    TestFeatureService ALL_FEATURES = ignored -> true;
+    Set<String> getAllSupportedFeatures();
+
+    TestFeatureService ALL_FEATURES = new TestFeatureService() {
```
Comment (thecoop) [inline test/framework/src/main/java/org/elasticsearch/test/rest/TestFeatureService.java:18]: Does it make sense to have this defined here, now the interface is more complicated? It's only used in ESRestTestCase, and the implementation of `getAllSupportedFeatures` references its use. So I think this should instead be the initial instance of `ESRestTestCase.testFeatureService`

Comment (ldematte) [reply test/framework/src/main/java/org/elasticsearch/test/rest/TestFeatureService.java:18]: That's a good point; I'll move it there.

File: qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/ParameterizedRollingUpgradeTestCase.java (hunk 1 above)
Comment (ldematte) [reply qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/ParameterizedRollingUpgradeTestCase.java:83]: It cannot be in `BeforeClass`, as it needs `testFeatureService` to be initialized - which for reasons happens in a `Before` method in the base class.
Besides, it follows the same pattern as the other initialization functions in the test, like `extractOldIndexVersion`.
My understanding is that everything is initialized in `Before` (keeping track of the upgrade "phase"), and when only when all tests from all phases are done we cleanup (that's why `AfterClass`)

File: test/framework/src/main/java/org/elasticsearch/test/rest/TestFeatureService.java (hunk 2 above)
Comment (ldematte) [reply test/framework/src/main/java/org/elasticsearch/test/rest/TestFeatureService.java:18]: I was also wondering if I should rename it to something like `UNINITIALIZED`, but decided to leave it as-is for now

**Review by thecoop — APPROVED** [review]

**Comment by ldematte** [comment]
Thanks Simon!

**Pull request merged**